### PR TITLE
Backend guide

### PR DESCRIPTION
**Pull request by jcrist1**
I feel like I'm getting a little bit annoying now, constantly pushing stuff 😅  but I wanted to get my own thoughts and experiences down building the kotlin backend, before I forget. I created a document which outlines how I would have done it starting again. There are definitely some things that I would have done differently, and I wanted to jot them down. In particular the `insta` tests, but also a more structured approach to initial development. @Manishearth is something like this  useful? If it's not really what you imagine for the backend guide, I'm happy to close, change, or move somewhere else.

### CODE REVIEW

**Comment by Manishearth** [comment]
Oh, thank you very much! I'll probably have some comments but it's great to have something like this, this book needs a lot of work especially in the new HIR world with all the new types.

**Comment by jcrist1** [comment]
😄 Cool! I got all of my thoughts out already and looks like I've finally configured my GH notifications to see comments on PRs I'm participating in. So if/when you've got some comments I can address them.

File: src/SUMMARY.md [hunk 1]
```diff
@@ -11,4 +11,5 @@
     - [Writeable](./writeable.md)
   - [Documentation](docs.md)
   - [Lifetimes](lifetimes.md)
-- [Backend developer guide](developer.md)
\ No newline at end of file
+- [Backend developer guide](developer.md)
```
Comment (Manishearth) [inline src/SUMMARY.md:14]: thought: should this file just be the toplevel guide?

File: src/backend_dev_setup.md [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+
+- [ ] **project structure**: You will need need to test your generated code so you should
+first set up a host language project. It should have all dependencies to be able to interface 
+with native code (or WASM).
+
+## Setting up Basic Code Generation in a Test
+
+Your backend should iterate over all [TypeDefs](https://docs.rs/diplomat_core/latest/diplomat_core/hir/enum.TypeDef.html)
+and generate the required code for these. A good starting point is to create a test for generating a simple opaque struct
+without any methods. Your backend should go in the tool create: create a module `tool/src/{backend}/mod.rs`
```
Comment (Manishearth) [inline src/backend_dev_setup.md]:
```suggestion
without any methods. Your backend should go in the tool crate: create a module `tool/src/{backend}/mod.rs` 
```

File: src/backend_dev_setup.md [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+
+You can now run 
+```sh
+cargo test -p diplomat-tool -- backend::test --nocapture
+```
+You should also have a generated snapshot `diplomat_tool__backend__test__opaque_gen.snap.new` 
+which you can use to pick up your generated code.
+
+## How to Generate the Dylib
+Now to actually test native methods you will need to create a dynamically linked library
```
Comment (Manishearth) [inline src/backend_dev_setup.md]: issue: not always, no. say that you will need to compile _some_ kind of library: a staticlib, dylib, or wasm lib. And then say here we'll make a dylib.

File: src/backend_dev_setup.md [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+- [ ] [**structs**](https://docs.rs/diplomat_core/0.7.0/diplomat_core/hir/struct.StructDef.html)
+- [ ] [**enums**](https://docs.rs/diplomat_core/0.7.0/diplomat_core/hir/struct.EnumDef.html)
+- [ ] [**writeable**](https://docs.rs/diplomat_core/0.7.0/diplomat_core/hir/enum.SuccessType.html#variant.Writeable)
+- [ ] [**slices**](https://docs.rs/diplomat_core/0.7.0/diplomat_core/hir/enum.Slice.html)
+  - [ ] primitive slices
+  - [ ] str slices
+  - [ ] owned slices should be kotlin arrays
+  - [ ] slices of strings
+  - [ ] strings
+- [ ] borrows. This is probably one of the trickiest things., as you need to ma
```
Comment (Manishearth) [inline src/backend_dev_setup.md]: issue: this is cut off

**Review by Manishearth — COMMENTED** [review]
seems okay, but could have more. still much better than what we had!

**Comment by jcrist1** [comment]
Thanks for the comments. I replaced the existing backend guide with this, but paraphrased some of the explanation that was there into this document, commented that there are many library options as a compilation target, and fixed the broken comment about borrows.

File: src/developr.md [hunk 5]
```diff
@@ -0,0 +1,235 @@
+# How to Set up a Minimal Backend
```
Comment (Manishearth) [inline src/developr.md]: nit: the file name is wrong

**Review by Manishearth — APPROVED** [review]

File: src/developr.md (hunk 5 above)
Comment (jcrist1) [reply src/developr.md]: fixed. Thanks for the review

**Review by Manishearth — APPROVED** [review]

**Pull request merged**